hw02_unpack_string/unpack.go

```
@@ -2,11 +2,64 @@ package hw02unpackstring
 
 import (
 	"errors"
+	"regexp"
+	"strconv"
+	"strings"
+	"unicode"
 )
 
 var ErrInvalidString = errors.New("invalid string")
 
-func Unpack(_ string) (string, error) {
-	// Place your code here.
-	return "", nil
+func Unpack(input string) (string, error) {
+	validation, err := validate(input)
+	if err != nil {
+		return "", err
+	}
+	if !validation {
+		return "", ErrInvalidString
+	}
+
+	var builder strings.Builder
+	runes := []rune(input)
+
+	for i := 0; i <= len(runes)-1; i++ {
+		char := string(runes[i])
+
+		if char == `\` {
+			builder.WriteRune(runes[i+1])
+			i++
+
+			continue
+		}
+
+		if unicode.IsDigit(runes[i]) {
+			repeatAmount, err := strconv.Atoi(char)
+			if err != nil {
+				return "", err
+			}
+			repeatLastChar(&builder, repeatAmount)
+			continue
+		}
+		builder.WriteString(char)
+	}
+
+	return builder.String(), nil
+}
+
+func repeatLastChar(builder *strings.Builder, repeatAmount int) {
+	currentString := []rune(builder.String())
+	lastChar := string(currentString[len(currentString)-1])
+
+	if repeatAmount > 0 {
+		builder.WriteString(strings.Repeat(lastChar, repeatAmount-1))
+	} else {
+		// if repeatAmount == 0 we need to cut previous char
+		builder.Reset()
+		builder.WriteString(string(currentString[:len(currentString)-1]))
+	}
+}
+
+func validate(input string) (bool, error) {
+	r, err := regexp.Compile(`(^\d)|([^\\]\d\d)|(\\[a-z]|\\[A-Z])|` + "(`)")
```

Регулярку лучше вынести в константы. + Если вернулась ошибка, то нет смысла вызывать функцию, стоит сразу возвращать ошибку (fail fast подход)